File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-com-5-4/CoalProduct.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class CoalProduct : ISerialized
    {
        public CoalProduct(XmlNode xmlNode)
        {
            XmlNode typeNode = xmlNode.SelectSingleNode("type");
            
            if (typeNode != null)
            {
                if (typeNode.Attributes["href"] != null || typeNode.Attributes["id"] != null) 
                {
                    if (typeNode.Attributes["id"] != null) 
                    {
                        typeIDRef_ = typeNode.Attributes["id"].Value;
                        CoalProductType ob = new CoalProductType(typeNode);
                        IDManager.SetID(typeIDRef_, ob);
                    }
                    else if (typeNode.Attributes["href"] != null)
                    {
                        typeIDRef_ = typeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        type_ = new CoalProductType(typeNode);
                    }
                }
                else
                {
                    type_ = new CoalProductType(typeNode);
                }
            }
            
        
            XmlNode coalProductSpecificationsNode = xmlNode.SelectSingleNode("coalProductSpecifications");
            
            if (coalProductSpecificationsNode != null)
            {
                if (coalProductSpecificationsNode.Attributes["href"] != null || coalProductSpecificationsNode.Attributes["id"] != null) 
                {
                    if (coalProductSpecificationsNode.Attributes["id"] != null) 
                    {
                        coalProductSpecificationsIDRef_ = coalProductSpecificationsNode.Attributes["id"].Value;
                        CoalProductSpecifications ob = new CoalProductSpecifications(coalProductSpecificationsNode);
                        IDManager.SetID(coalProductSpecificationsIDRef_, ob);
                    }
                    else if (coalProductSpecificationsNode.Attributes["href"] != null)
                    {
                        coalProductSpecificationsIDRef_ = coalProductSpecificationsNode.Attributes["href"].Value;
                    }
                    else
                    {
                        coalProductSpecifications_ = new CoalProductSpecifications(coalProductSpecificationsNode);
                    }
                }
                else
                {
                    coalProductSpecifications_ = new CoalProductSpecifications(coalProductSpecificationsNode);
                }
            }
            
        
            XmlNodeList sourceNodeList = xmlNode.SelectNodes("source");
            
            if (sourceNodeList != null)
            {
                this.source_ = new List<CoalProductSource>();
                foreach (XmlNode item in sourceNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            sourceIDRef_ = item.Attributes["id"].Value;
                            source_.Add(new CoalProductSource(item));
                            IDManager.SetID(sourceIDRef_, source_[source_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            sourceIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        source_.Add(new CoalProductSource(item));
                        }
                    }
                    else
                    {
                        source_.Add(new CoalProductSource(item));
                    }
                }
            }
            
        
            XmlNode sCoTASpecificationsNode = xmlNode.SelectSingleNode("sCoTASpecifications");
            
            if (sCoTASpecificationsNode != null)
            {
                if (sCoTASpecificationsNode.Attributes["href"] != null || sCoTASpecificationsNode.Attributes["id"] != null) 
                {
                    if (sCoTASpecificationsNode.Attributes["id"] != null) 
                    {
                        sCoTASpecificationsIDRef_ = sCoTASpecificationsNode.Attributes["id"].Value;
                        XsdTypeBoolean ob = new XsdTypeBoolean(sCoTASpecificationsNode);
                        IDManager.SetID(sCoTASpecificationsIDRef_, ob);
                    }
                    else if (sCoTASpecificationsNode.Attributes["href"] != null)
                    {
                        sCoTASpecificationsIDRef_ = sCoTASpecificationsNode.Attributes["href"].Value;
                    }
                    else
                    {
                        sCoTASpecifications_ = new XsdTypeBoolean(sCoTASpecificationsNode);
                    }
                }
                else
                {
                    sCoTASpecifications_ = new XsdTypeBoolean(sCoTASpecificationsNode);
                }
            }
            
        
            XmlNode btuQualityAdjustmentNode = xmlNode.SelectSingleNode("btuQualityAdjustment");
            
            if (btuQualityAdjustmentNode != null)
            {
                if (btuQualityAdjustmentNode.Attributes["href"] != null || btuQualityAdjustmentNode.Attributes["id"] != null) 
                {
                    if (btuQualityAdjustmentNode.Attributes["id"] != null) 
                    {
                        btuQualityAdjustmentIDRef_ = btuQualityAdjustmentNode.Attributes["id"].Value;
                        CoalQualityAdjustments ob = new CoalQualityAdjustments(btuQualityAdjustmentNode);
                        IDManager.SetID(btuQualityAdjustmentIDRef_, ob);
                    }
                    else if (btuQualityAdjustmentNode.Attributes["href"] != null)
                    {
                        btuQualityAdjustmentIDRef_ = btuQualityAdjustmentNode.Attributes["href"].Value;
                    }
                    else
                    {
                        btuQualityAdjustment_ = new CoalQualityAdjustments(btuQualityAdjustmentNode);
                    }
                }
                else
                {
                    btuQualityAdjustment_ = new CoalQualityAdjustments(btuQualityAdjustmentNode);
                }
            }
            
        
            XmlNode so2QualityAdjustmentNode = xmlNode.SelectSingleNode("so2QualityAdjustment");
            
            if (so2QualityAdjustmentNode != null)
            {
                if (so2QualityAdjustmentNode.Attributes["href"] != null || so2QualityAdjustmentNode.Attributes["id"] != null) 
                {
                    if (so2QualityAdjustmentNode.Attributes["id"] != null) 
                    {
                        so2QualityAdjustmentIDRef_ = so2QualityAdjustmentNode.Attributes["id"].Value;
                        CoalQualityAdjustments ob = new CoalQualityAdjustments(so2QualityAdjustmentNode);
                        IDManager.SetID(so2QualityAdjustmentIDRef_, ob);
                    }
                    else if (so2QualityAdjustmentNode.Attributes["href"] != null)
                    {
                        so2QualityAdjustmentIDRef_ = so2QualityAdjustmentNode.Attributes["href"].Value;
                    }
                    else
                    {
                        so2QualityAdjustment_ = new CoalQualityAdjustments(so2QualityAdjustmentNode);
                    }
                }
                else
                {
                    so2QualityAdjustment_ = new CoalQualityAdjustments(so2QualityAdjustmentNode);
                }
            }
            
        
        }
        
    
        #region Type_
        private CoalProductType type_;
        public CoalProductType Type_
        {
            get
            {
                if (this.type_ != null)
                {
                    return this.type_; 
                }
                else if (this.typeIDRef_ != null)
                {
                    type_ = IDManager.getID(typeIDRef_) as CoalProductType;
                    return this.type_; 
                }
                else
                {
                      return this.type_; 
                }
            }
            set
            {
                if (this.type_ != value)
                {
                    this.type_ = value;
                }
            }
        }
        #endregion
        
        public string typeIDRef_ { get; set; }
        #region CoalProductSpecifications_
        private CoalProductSpecifications coalProductSpecifications_;
        public CoalProductSpecifications CoalProductSpecifications_
        {
            get
            {
                if (this.coalProductSpecifications_ != null)
                {
                    return this.coalProductSpecifications_; 
                }
                else if (this.coalProductSpecificationsIDRef_ != null)
                {
                    coalProductSpecifications_ = IDManager.getID(coalProductSpecificationsIDRef_) as CoalProductSpecifications;
                    return this.coalProductSpecifications_; 
                }
                else
                {
                      return this.coalProductSpecifications_; 
                }
            }
            set
            {
                if (this.coalProductSpecifications_ != value)
                {
                    this.coalProductSpecifications_ = value;
                }
            }
        }
        #endregion
        
        public string coalProductSpecificationsIDRef_ { get; set; }
        #region Source_
        private List<CoalProductSource> source_;
        public List<CoalProductSource> Source_
        {
            get
            {
                if (this.source_ != null)
                {
                    return this.source_; 
                }
                else if (this.sourceIDRef_ != null)
                {
                    return this.source_; 
                }
                else
                {
                      return this.source_; 
                }
            }
            set
            {
                if (this.source_ != value)
                {
                    this.source_ = value;
                }
            }
        }
        #endregion
        
        public string sourceIDRef_ { get; set; }
        #region SCoTASpecifications_
        private XsdTypeBoolean sCoTASpecifications_;
        public XsdTypeBoolean SCoTASpecifications_
        {
            get
            {
                if (this.sCoTASpecifications_ != null)
                {
                    return this.sCoTASpecifications_; 
                }
                else if (this.sCoTASpecificationsIDRef_ != null)
                {
                    sCoTASpecifications_ = IDManager.getID(sCoTASpecificationsIDRef_) as XsdTypeBoolean;
                    return this.sCoTASpecifications_; 
                }
                else
                {
                      return this.sCoTASpecifications_; 
                }
            }
            set
            {
                if (this.sCoTASpecifications_ != value)
                {
                    this.sCoTASpecifications_ = value;
                }
            }
        }
        #endregion
        
        public string sCoTASpecificationsIDRef_ { get; set; }
        #region BtuQualityAdjustment_
        private CoalQualityAdjustments btuQualityAdjustment_;
        public CoalQualityAdjustments BtuQualityAdjustment_
        {
            get
            {
                if (this.btuQualityAdjustment_ != null)
                {
                    return this.btuQualityAdjustment_; 
                }
                else if (this.btuQualityAdjustmentIDRef_ != null)
                {
                    btuQualityAdjustment_ = IDManager.getID(btuQualityAdjustmentIDRef_) as CoalQualityAdjustments;
                    return this.btuQualityAdjustment_; 
                }
                else
                {
                      return this.btuQualityAdjustment_; 
                }
            }
            set
            {
                if (this.btuQualityAdjustment_ != value)
                {
                    this.btuQualityAdjustment_ = value;
                }
            }
        }
        #endregion
        
        public string btuQualityAdjustmentIDRef_ { get; set; }
        #region So2QualityAdjustment_
        private CoalQualityAdjustments so2QualityAdjustment_;
        public CoalQualityAdjustments So2QualityAdjustment_
        {
            get
            {
                if (this.so2QualityAdjustment_ != null)
                {
                    return this.so2QualityAdjustment_; 
                }
                else if (this.so2QualityAdjustmentIDRef_ != null)
                {
                    so2QualityAdjustment_ = IDManager.getID(so2QualityAdjustmentIDRef_) as CoalQualityAdjustments;
                    return this.so2QualityAdjustment_; 
                }
                else
                {
                      return this.so2QualityAdjustment_; 
                }
            }
            set
            {
                if (this.so2QualityAdjustment_ != value)
                {
                    this.so2QualityAdjustment_ = value;
                }
            }
        }
        #endregion
        
        public string so2QualityAdjustmentIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-com-5-4/CoalProduct.cpp
// CoalProduct.cpp 
#include "CoalProduct.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

CoalProduct::CoalProduct(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //typeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* typeNode = xmlNode->FirstChildElement("type");

   if(typeNode){typeIsNull_ = false;}
   else{typeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- typeNode , address : " << typeNode << std::endl;
   #endif
   if(typeNode)
   {
      if(typeNode->Attribute("href") || typeNode->Attribute("id"))
      {
          if(typeNode->Attribute("id"))
          {
             typeIDRef_ = typeNode->Attribute("id");
             type_ = boost::shared_ptr<CoalProductType>(new CoalProductType(typeNode));
             type_->setID(typeIDRef_);
             IDManager::instance().setID(typeIDRef_,type_);
          }
          else if(typeNode->Attribute("href")) { typeIDRef_ = typeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { type_ = boost::shared_ptr<CoalProductType>(new CoalProductType(typeNode));}
   }

   //coalProductSpecificationsNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* coalProductSpecificationsNode = xmlNode->FirstChildElement("coalProductSpecifications");

   if(coalProductSpecificationsNode){coalProductSpecificationsIsNull_ = false;}
   else{coalProductSpecificationsIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- coalProductSpecificationsNode , address : " << coalProductSpecificationsNode << std::endl;
   #endif
   if(coalProductSpecificationsNode)
   {
      if(coalProductSpecificationsNode->Attribute("href") || coalProductSpecificationsNode->Attribute("id"))
      {
          if(coalProductSpecificationsNode->Attribute("id"))
          {
             coalProductSpecificationsIDRef_ = coalProductSpecificationsNode->Attribute("id");
             coalProductSpecifications_ = boost::shared_ptr<CoalProductSpecifications>(new CoalProductSpecifications(coalProductSpecificationsNode));
             coalProductSpecifications_->setID(coalProductSpecificationsIDRef_);
             IDManager::instance().setID(coalProductSpecificationsIDRef_,coalProductSpecifications_);
          }
          else if(coalProductSpecificationsNode->Attribute("href")) { coalProductSpecificationsIDRef_ = coalProductSpecificationsNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { coalProductSpecifications_ = boost::shared_ptr<CoalProductSpecifications>(new CoalProductSpecifications(coalProductSpecificationsNode));}
   }

   //sourceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* sourceNode = xmlNode->FirstChildElement("source");

   if(sourceNode){sourceIsNull_ = false;}
   else{sourceIsNull_ = true;}

   if(sourceNode)
   {
      for(sourceNode; sourceNode; sourceNode = sourceNode->NextSiblingElement("source")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- sourceNode , address : " << sourceNode << std::endl;
          #endif
          if(sourceNode->Attribute("href") || sourceNode->Attribute("id"))
          {
              if(sourceNode->Attribute("id"))
              {
                  sourceIDRef_ = sourceNode->Attribute("id");
                  source_.push_back(boost::shared_ptr<CoalProductSource>(new CoalProductSource(sourceNode)));
                  source_.back()->setID(sourceIDRef_);
                  IDManager::instance().setID(sourceIDRef_, source_.back());
              }
              else if(sourceNode->Attribute("href")) { sourceIDRef_ = sourceNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { source_.push_back(boost::shared_ptr<CoalProductSource>(new CoalProductSource(sourceNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- sourceNode , address : " << sourceNode << std::endl;
       #endif
   }

   //sCoTASpecificationsNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* sCoTASpecificationsNode = xmlNode->FirstChildElement("sCoTASpecifications");

   if(sCoTASpecificationsNode){sCoTASpecificationsIsNull_ = false;}
   else{sCoTASpecificationsIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- sCoTASpecificationsNode , address : " << sCoTASpecificationsNode << std::endl;
   #endif
   if(sCoTASpecificationsNode)
   {
      if(sCoTASpecificationsNode->Attribute("href") || sCoTASpecificationsNode->Attribute("id"))
      {
          if(sCoTASpecificationsNode->Attribute("id"))
          {
             sCoTASpecificationsIDRef_ = sCoTASpecificationsNode->Attribute("id");
             sCoTASpecifications_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(sCoTASpecificationsNode));
             sCoTASpecifications_->setID(sCoTASpecificationsIDRef_);
             IDManager::instance().setID(sCoTASpecificationsIDRef_,sCoTASpecifications_);
          }
          else if(sCoTASpecificationsNode->Attribute("href")) { sCoTASpecificationsIDRef_ = sCoTASpecificationsNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { sCoTASpecifications_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(sCoTASpecificationsNode));}
   }

   //btuQualityAdjustmentNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* btuQualityAdjustmentNode = xmlNode->FirstChildElement("btuQualityAdjustment");

   if(btuQualityAdjustmentNode){btuQualityAdjustmentIsNull_ = false;}
   else{btuQualityAdjustmentIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- btuQualityAdjustmentNode , address : " << btuQualityAdjustmentNode << std::endl;
   #endif
   if(btuQualityAdjustmentNode)
   {
      if(btuQualityAdjustmentNode->Attribute("href") || btuQualityAdjustmentNode->Attribute("id"))
      {
          if(btuQualityAdjustmentNode->Attribute("id"))
          {
             btuQualityAdjustmentIDRef_ = btuQualityAdjustmentNode->Attribute("id");
             btuQualityAdjustment_ = boost::shared_ptr<CoalQualityAdjustments>(new CoalQualityAdjustments(btuQualityAdjustmentNode));
             btuQualityAdjustment_->setID(btuQualityAdjustmentIDRef_);
             IDManager::instance().setID(btuQualityAdjustmentIDRef_,btuQualityAdjustment_);
          }
          else if(btuQualityAdjustmentNode->Attribute("href")) { btuQualityAdjustmentIDRef_ = btuQualityAdjustmentNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { btuQualityAdjustment_ = boost::shared_ptr<CoalQualityAdjustments>(new CoalQualityAdjustments(btuQualityAdjustmentNode));}
   }

   //so2QualityAdjustmentNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* so2QualityAdjustmentNode = xmlNode->FirstChildElement("so2QualityAdjustment");

   if(so2QualityAdjustmentNode){so2QualityAdjustmentIsNull_ = false;}
   else{so2QualityAdjustmentIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- so2QualityAdjustmentNode , address : " << so2QualityAdjustmentNode << std::endl;
   #endif
   if(so2QualityAdjustmentNode)
   {
      if(so2QualityAdjustmentNode->Attribute("href") || so2QualityAdjustmentNode->Attribute("id"))
      {
          if(so2QualityAdjustmentNode->Attribute("id"))
          {
             so2QualityAdjustmentIDRef_ = so2QualityAdjustmentNode->Attribute("id");
             so2QualityAdjustment_ = boost::shared_ptr<CoalQualityAdjustments>(new CoalQualityAdjustments(so2QualityAdjustmentNode));
             so2QualityAdjustment_->setID(so2QualityAdjustmentIDRef_);
             IDManager::instance().setID(so2QualityAdjustmentIDRef_,so2QualityAdjustment_);
          }
          else if(so2QualityAdjustmentNode->Attribute("href")) { so2QualityAdjustmentIDRef_ = so2QualityAdjustmentNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { so2QualityAdjustment_ = boost::shared_ptr<CoalQualityAdjustments>(new CoalQualityAdjustments(so2QualityAdjustmentNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<CoalProductType> CoalProduct::getType()
{
   if(!this->typeIsNull_){
        if(typeIDRef_ != ""){
             return boost::shared_static_cast<CoalProductType>(IDManager::instance().getID(typeIDRef_));
        }else{
             return this->type_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CoalProductType>();
   }
}
boost::shared_ptr<CoalProductSpecifications> CoalProduct::getCoalProductSpecifications()
{
   if(!this->coalProductSpecificationsIsNull_){
        if(coalProductSpecificationsIDRef_ != ""){
             return boost::shared_static_cast<CoalProductSpecifications>(IDManager::instance().getID(coalProductSpecificationsIDRef_));
        }else{
             return this->coalProductSpecifications_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CoalProductSpecifications>();
   }
}
std::vector<boost::shared_ptr<CoalProductSource>> CoalProduct::getSource()
{
   if(!this->sourceIsNull_){
        if(sourceIDRef_ != ""){
             return this->source_;
        }else{
             return this->source_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<CoalProductSource>>();
   }
}
boost::shared_ptr<XsdTypeBoolean> CoalProduct::getSCoTASpecifications()
{
   if(!this->sCoTASpecificationsIsNull_){
        if(sCoTASpecificationsIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeBoolean>(IDManager::instance().getID(sCoTASpecificationsIDRef_));
        }else{
             return this->sCoTASpecifications_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeBoolean>();
   }
}
boost::shared_ptr<CoalQualityAdjustments> CoalProduct::getBtuQualityAdjustment()
{
   if(!this->btuQualityAdjustmentIsNull_){
        if(btuQualityAdjustmentIDRef_ != ""){
             return boost::shared_static_cast<CoalQualityAdjustments>(IDManager::instance().getID(btuQualityAdjustmentIDRef_));
        }else{
             return this->btuQualityAdjustment_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CoalQualityAdjustments>();
   }
}
boost::shared_ptr<CoalQualityAdjustments> CoalProduct::getSo2QualityAdjustment()
{
   if(!this->so2QualityAdjustmentIsNull_){
        if(so2QualityAdjustmentIDRef_ != ""){
             return boost::shared_static_cast<CoalQualityAdjustments>(IDManager::instance().getID(so2QualityAdjustmentIDRef_));
        }else{
             return this->so2QualityAdjustment_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CoalQualityAdjustments>();
   }
}
}

